package management: straight -> elpaca

balacij opened the pull request:
I'm intentionally breaking up the commits into smaller bits for reviewability. I will add a few notes on the changes as well by reviewing the PR.

Regarding commits and their messages: my thoughts are you could review this PR, suggest changes/edit it, and then we can squash and rewrite the messages.

EDIT: Just to be clear, I don't think this is merge-ready yet. There's work to be done to clean it up.

balacij committed:
Add (unused) module for elpaca-based package management

balacij committed:
Move straight.el extras to new core module: basics.el

balacij committed:
Switch package manager: straight -> elpaca

balacij committed:
Update theme-related modules for elpaca-specifics

balacij committed:
Update core-related modules for elpaca

balacij committed:
Update editor-related modules for elpaca

balacij committed:
Update lang-related modules for elpaca

balacij committed:
Update tools-related modules for elpaca

balacij commented on modules/core/elpaca.el:
```diff
@@ -0,0 +1,53 @@
+;;; core/elpaca --- DESCRIPTION -*- no-byte-compile: t; lexical-binding: t; -*-
+
+;; BEGIN ELPACA SNIPPET: https://github.com/progfolio/elpaca?tab=readme-ov-file#installer
```
As you can see from this, this is just a snippet that comes directly from that link.

TOTBWF replied:
Made some tweaks to resolve some issues I had with `elpaca-core-date`

balacij commented on modules/editor/imenu.el:
```diff
@@ -5,10 +5,10 @@
+  :ensure nil
   :bind
   (:map meow-leader-keymap
-	("i" ("imenu" . imenu))))
```
Maybe the load order is bad? This is not working anymore.

TOTBWF replied:
Probably! It's currently a disaster :)

balacij commented on modules/editor/project.el:
```diff
@@ -22,12 +22,13 @@
-    :host github
-    :repo "aurtzy/disproject")
+  :ensure
+    '(disproject
```
Do you prefer extra indentation for things like this?

TOTBWF replied:
I'd like it if that was on the same line.

Do these lists have to be quoted? That's kind of annoying if so.

balacij replied:
disproject did not need to be quoted, and fixed the indendentation in 6e2b733. Thanks!

balacij commented on modules/lang/elisp.el:
```diff
@@ -8,7 +8,8 @@
 (use-package elisp-mode
-  :straight nil
+  :ensure nil
+  :after company
```
Surprisingly gave me a headache! I was getting a warning on init about functions being unable to be defined because of a missing symbol: company-mode. With async loading, the load order needs a bit more attention indeed, but the majority of your config worked just fine (at least from what I've seen -- I haven't tested all the features yet).

TOTBWF replied:
Company is a pain, and should definitely be put quite early in the load order. I do have some hacks at https://github.com/TOTBWF/mumacs/blob/b801381684ba30ada6e4bca8287730f8606c6cdb/modules/editor/company.el#L23 for handling company `use-package` integration, so might be worth looking at.

balacij commented on modules/lang/idris.el:
```diff
@@ -4,7 +4,7 @@
 ;;; Code:
 (use-package idris2-mode
-  :straight (idris2-mode :type git :host github :repo "idris-community/idris2-mode"))
+  :ensure (idris2-mode :type git :host github :repo "idris-community/idris2-mode"))
```
Untested because it's not imported by init.el.

balacij committed:
fixup! lang/latex: no longer using straight

balacij commented on modules/lang/lean.el:
```diff
@@ -6,7 +6,7 @@
 (use-package lean4-mode
   :requires polymode markdown-mode
-  :straight (lean4-mode
+  :ensure (lean4-mode
```
Untested as well, for similar reasons.

balacij commented on modules/lang/markdown.el:
```diff
@@ -5,7 +5,7 @@
 
 (use-package markdown-mode
-  :straight nil
+  :ensure nil
```
I believe that markdown-mode is now imported twice (agda.el and here) -- do you want them reconciled?

TOTBWF replied:
This is detritus from me trying to fix the annoying behavior of `markdown-mode`, don't worry about reconciling.

I'll also note that multiple `use-package` blocks don't cause the package to get loaded twice. This confused me quite a bit when I started learning how it worked: would recommend poking around at the result of `pp-macroexpand-last-sexp` to get a feel for what is going on behind the scenes.

As an aside, I've been thinking about writing my own version of `use-package` that's a bit simpler; would be happy to discuss.

balacij commented on modules/lang/ocaml.el:
```diff
@@ -45,12 +52,13 @@
   :bind
   (:map opam-switch-mode-map
-	("C-c C-w" ("set switch" . opam-switch-set-switch))))
+	("C-c C-w" . opam-switch-set-switch)))
```
Another broken feature.

TOTBWF replied:
Huh, very odd. This should add `which-key` hints, but I'm not sure when it broke!

balacij replied:
It looks like [`:bind` expands to `:init` with `bind-key`s](https://github.com/jwiegley/use-package?tab=readme-ov-file#key-binding), which, as we realized yesterday with `vertico`, does not play well with elpaca.

balacij commented on modules/lang/tlaplus.el:
```diff
@@ -19,12 +19,12 @@
 
 (use-package tla-mode
   :mode "\\.tla\\'"
-    :straight
```
Untested again, because it is not used by init.el

TOTBWF replied:
This file should be banished :)

balacij committed:
fixup! tools/ripgrep: use elpaca

balacij committed:
fixup! early-init: change a comment that I don't fully understand :^)

balacij committed:
fixup! tools/org: remove straight references

balacij committed:
fixup! core/*: remove straight references

balacij commented on early-init.el:
```diff
@@ -4,7 +4,7 @@
 
 ;;; Code:
 
-;; We use `straight.el' for package management, so we disable
+;; We use `elpaca.el' for package management, so we disable
 ;; `package.el' at startup to avoid loading packages twice.
 (setq package-enable-at-startup nil)
```
I'm editing the comment, but I'm not sure about it. Wouldn't it be good to set it to `t`?

TOTBWF replied:
This has to do with another built-in packaging system for emacs called `package.el` that leaves a lot to be desired. We don't use it anywhere, so it off is safe, and really speeds up load times.

TOTBWF commented:
Tried this out and ran into an issue building `transient`: something about
```
compat installed version (30 0 9999) lower than min required 30.1.0.0 
```

I vaguely remember you had a solution for this? I am running emacs HEAD, but haven't built a new copy in a few months so it could be that.

balacij committed:
fix: replace `:init' for `vertico-mode' with `elpaca-after-init-hook'

Using :init seems to result in some odd load behaviour. I suspect
that we want to be using elpaca-after-init-hook 99% of the time.

Co-authored-by: Reed Mullanix <reedmullanix@gmail.com>

balacij committed:
fixup! fix tools/vterm

balacij committed:
fixup! spell-fu

balacij committed:
fixup! path environment variables

balacij committed:
fixup! reorder org <-> magit dependency

balacij committed:
fixup! core/help: ensure elisp-demos is installed and disable imenu (where does it come from?)

balacij committed:
fixup! orgit, orgit-forge dependencies and magit keymap

Co-authored-by: Reed Mullanix <reedmullanix@gmail.com>

balacij committed:
feat: enable debugging information on error

balacij committed:
feat: add a macro to define meow-leader-keymaps quickly

Co-authored-by: Reed Mullanix <reedmullanix@gmail.com>

balacij committed:
fixup! lang/latex hooks

balacij commented:
You might be able to add `(elpaca compat)` before `(elpaca transient)` in `project.el`. If that doesn't work, you can try deleting `(elpaca transient)`.

My issue was similar: the shipped version of transient was too low for `disproject` to build. That was why I added the `transient` to the list of dependencies that elpaca manages

For reference, I'm using Emacs 30.1.

TOTBWF commented:
Upgraded my emacs and the problem resolved itself :)

balacij committed:
fixup! project.el: add elpaca require, remove quote on disproject dependency, and dedent disproject dependency

balacij committed:
fix: tua-reg-mode-hook typo in ocaml.el

balacij commented:
Ah, perfect!

balacij committed:
fix: use emacsattic mirror as iman's source

Co-authored-by: Reed Mullanix <reedmullanix@gmail.com>

TOTBWF committed:
feat: robust macros for defining leader/keypad entries

TOTBWF committed:
fix: use `define-leader' and `define-keypad'

TOTBWF committed:
fix: rename references to `meow-leader-keymap'

TOTBWF committed:
fix: point to fork of `orgit-forge'

TOTBWF committed:
fix: set `elpaca-core-date' even on nix-built emacs

TOTBWF committed:
chore: factor out :advice into its own file

TOTBWF committed:
chore: set use-package opts before elpaca load

TOTBWF committed:
feat: use flymake-elpaca for elisp

TOTBWF committed:
fix: fix silly error in `core/meow`

TOTBWF committed:
fix: minor typo

TOTBWF committed:
fix: missing :ensure

TOTBWF committed:
fix: check that meow-keypad-start-keys is bound in define-keymap

This silences the byte compiler, and leads to a more robust macro
in the long term.

TOTBWF committed:
fix: dont run `elpaca-core-date' hack when byte compiling

TOTBWF committed:
fix: formatting

TOTBWF committed:
fix: some missing :ensures

balacij commented on modules/editor/project.el:
```diff
@@ -20,10 +20,14 @@
     (file-relative-name filename (project-root (or project (project-current))))))
 
-(elpaca transient) ;; Make sure we're using an up-to-date copy of transient?
+;; HACK: This belongs elsewhere.
+;; Make sure that we have an up-to-date copy of transient.
+(elpaca transient)
+
 (use-package disproject
```
`package-overlays.el`?

TOTBWF replied:
Could just move it to core/transient?

balacij replied:
Just moved it :)

balacij commented on modules/tools/direnv.el:
```diff
@@ -3,9 +3,12 @@
 ;;; Commentary:
 
 ;;; Code:
+(require 'core/elpaca)
+
 (use-package envrc
+  :ensure t
   :config
-  (after-init-hook . envrc-global-mode))
+  (after-elpaca-init-hook . envrc-global-mode))
 
 (provide 'tools/direnv)
```
I'm surprised at the number of `:ensure t`s I missed! Thank you!

balacij commented:
Re: my commits: feel free to squash/rewrite mine!

balacij commented on modules/editor/compilation.el:
```diff
@@ -6,13 +6,14 @@
 
 ;; Replace ANSI color escape codes with faces.
 (use-package ansi-color
+  :ensure t
```
I think this package is a built-in?

I'm getting:
```
Unable to determine repo dir: (error "Cannot determine URL from recipe: (:source nil :protocol https :inherit t :depth treeless :package \"ansi-color\")")
```

Changing to `:ensure nil` gives me no warnings, but I'm not sure if I'm just not running into the 'right' errors.

TOTBWF replied:
Oops :)

balacij committed:
fix: don't tell elpaca to try downloading ansi-color

balacij committed:
fix: ensure elpaca installs cdlatex

balacij committed:
chore: move transient dep require to core/transient

Suggested-by: Reed Mullanix <reedmullanix@gmail.com>

balacij committed:
fix: include `/elpa/' in .gitignore

balacij commented:
One warning I'm getting on start-up is:
```
 ⛔ Warning (emacs): define-keypad: ‘meow-keypad-start-keys’ not bound.
```

TOTBWF committed:
ci: first attempt at CI

TOTBWF commented:
Hmmm, that should be fixed by https://github.com/TOTBWF/mumacs/pull/1/commits/689527df2fd691efa293a1952542d3828ad9b8eb, odd!

TOTBWF commented:
Going to see if I can get CI working: for some reason I can't do it on this branch?

balacij commented:
Hmm, odd, I imagine it's a security policy issue? Not sure, but it looks like you're already working on `main` :)

TOTBWF commented:
Yeah, I could spend an hour reading github docs or just go full 🤠; decided that was easier :P

TOTBWF commented:
Huh, looks like it did actually run, just on your repo! Once I get it working reliably on `main` we can merge and do some final tweaks. If that passes without too many warnings, I say we just merge and clean up as we go.

TOTBWF committed:
Merge branch 'main' into elpaca

TOTBWF committed:
ci: fix formatting of messages

TOTBWF committed:
ci: print emacs errors better

TOTBWF committed:
fix: more robust handling of spell-fu use-package keyword

TOTBWF committed:
fix: put warning inside of if for define-keypad

TOTBWF committed:
chore: requires for lang/agda

TOTBWF committed:
ci: fix yaml formatting

TOTBWF committed:
ci: try pulling loading logic into ci.el

TOTBWF committed:
fix: remove boundp checks on spelling

TOTBWF committed:
ci: modify printing of emacs errors again

TOTBWF committed:
chore: remove rzk for now

TOTBWF committed:
fix: replace deprecated if-let with if-let*

TOTBWF committed:
chore: wrap load inside of group for gh actions

TOTBWF committed:
ci: warning prefix function

TOTBWF committed:
ci: add trailing :: to warning prefix function

TOTBWF committed:
chore: clean up `tools/org.el'

TOTBWF committed:
doc: document the `agda-mode' hacks

TOTBWF committed:
snippets: snippets for latex

TOTBWF committed:
chore: gitignore

TOTBWF committed:
ci: fix warning prefix function

TOTBWF committed:
ci: dont use cons cell

TOTBWF committed:
chore: imenu support for use-package

TOTBWF committed:
feat: flymake keybindings

balacij commented:
Re: `.gitignore`: we can switch to a negative `.gitignore`. Packages seem to create a lot of junk in the configuration directory of Emacs.

e.g.,
```gitignore
*
!init.el
!early-init.el
!/modules/core/
!/modules/core/*
!/modules/lang/
!/modules/lang/*
```

TOTBWF committed:
chore: clean up :advice keyword

TOTBWF committed:
chore: clean up latex + org integration

TOTBWF commented:
Seems like a good call

TOTBWF committed:
chore: negative .gitignore

TOTBWF committed:
chore: disable menu decorations in early-init

TOTBWF committed:
chore: pull most UI tweaks into early-init

TOTBWF committed:
fix: fix typo in elpaca-after-init-hook

TOTBWF committed:
chore: remove redundant camel-case helper

s.el provides `s-lower-camel-case' and `s-upper-camel-case'.

TOTBWF committed:
chore: shorter doc-string on compilation start advice

TOTBWF committed:
chore: remove :elpaca keyword

TOTBWF committed:
chore: refactor org a bit

I previously had `org-agenda' in its own use-package block,
even though it comes from `org'. This is not a good idea with
`elpaca': using `:ensure nil' will cause it to get loaded during
init time, not when elpaca processes its queues.

TOTBWF committed:
chore: remove :elpaca keyword

TOTBWF committed:
hack: silence warnings in lang/agda

TOTBWF committed:
fix: fix direnv

TOTBWF committed:
fix: better hacks for agda/markdown

TOTBWF commented:
Sure we will find some minor things, but this seems good to merge for now!

balacij commented:
Sounds good to me :)
